### PR TITLE
Minimize item

### LINKED ISSUE #307: creator/viewer should be able to minimize the item popup to view the video screen behind and maximize the question again

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #307 

## Test Plan

- [x] Test Responsiveness
    - [x] Laptop (1200px)
    - [x] Tablet (760px)
    - [x] Phone (320px)
- [x] Cross-Browser Testing
    - [x] Chrome
    - [x] Firefox
    - [x] Safari
- [x] Local Language Support
- [x] Tested locally
- [ ] Tested on staging
- [ ] Tested on production

### CODE REVIEW

File: src/assets/locales/hi.js [hunk 1]
```diff
@@ -29,6 +29,8 @@ export default {
         published: "परिवर्तन प्रकाशित करें",
       },
       add_question: "नया सवाल बनाएं",
+      show_video: "वीडियो देखें",
+      show_item: "वस्तु देखें",
```
Comment (dalmia) [inline src/assets/locales/hi.js:33]: show question

File: src/assets/locales/en.js [hunk 2]
```diff
@@ -29,6 +29,8 @@ export default {
         published: "Publish Changes",
       },
       add_question: "Add a question",
+      show_video: "Show Video",
+      show_item: "Show Item",
```
Comment (dalmia) [inline src/assets/locales/en.js:33]: show question

File: src/components/Items/Question/Footer.vue [hunk 3]
```diff
@@ -23,6 +23,7 @@
         :titleConfig="fullscreenButtonTitleConfig"
         :buttonClass="fullscreenButtonClass"
         @click="toggleFullscreen"
+        class="btn"
```
Comment (dalmia) [inline src/components/Items/Question/Footer.vue:26]: why this? why is this name for the class appropriate?

File: src/components/Items/Question/Footer.vue [hunk 4]
```diff
@@ -181,7 +182,7 @@ export default {
 };
 </script>
 
-<style scoped>
+<style>
```
Comment (dalmia) [inline src/components/Items/Question/Footer.vue:185]: why this?

File: src/components/UI/Buttons/IconButton.vue [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <button
     type="button"
     :class="buttonClass"
-    class="flex justify-center items-center transition ease-in duration-200 text-center text-base font-semibold focus:shadow-none focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed"
+    class="flex justify-center items-center transition ease-in duration-200 text-center text-base font-semibold focus:shadow-none focus:outline-none disabled:opacity-50 disabled:cursor-not-allowed"
```
Comment (dalmia) [inline src/components/UI/Buttons/IconButton.vue:5]: why?

File: src/components/Items/Question/Header.vue [hunk 6]
```diff
@@ -1,25 +1,52 @@
...
-  <div class="flex w-full bg-white justify-end p-1 pr-3 md:pr-6">
+  <div class="flex w-full bg-white justify-end p-1 pr-3 md:pr-6 space-x-2">
     <!-- skip button -->
     <icon-button
-      :iconConfig="skipButtonIconConfig"
+      :titleConfig="skipButtonTitleConfig"
       @click="skipClicked"
       :class="{ hidden: isAnswerSubmitted }"
+      :buttonClass="skipButtonClass"
+      class="btn"
```
Comment (dalmia) [inline src/components/Items/Question/Header.vue:9]: what is this class btn?

File: src/pages/Editor.vue [hunk 7]
```diff
@@ -1000,3 +1036,100 @@ export default {
...
+}
+
+.modalGrowShrink-leave {
+  @include modalScale(0);
+}
+
+.modalGrowShrink-enter {
+  @include modalScale(1);
+}
+</style>
```
Comment (dalmia) [inline src/pages/Editor.vue:1135]: need an explanation for this.

File: src/pages/Player.vue [hunk 8]
```diff
@@ -26,21 +33,42 @@
...
+        :class="[minimizeButtonPositionClass, 'absolute z-20 btn']"
+        id="maximizeButton"
+      >
+      </icon-button>
+      <!-- transition for minimizing/maximizing item modal -->
+      <transition
+        name="modalGrowShrink"
+        enter-active-class="grow"
+        leave-active-class="shrink"
+      >
```
Comment (dalmia) [inline src/pages/Player.vue:51]: need an explanation for this.

File: src/pages/Player.vue [hunk 9]
```diff
@@ -274,6 +336,18 @@ export default {
...
+    toggleMinimize() {
+      // toggle the minimized state of the modal
+      this.isModalMinimized = !this.isModalMinimized;
+      if (this.isFullscreen) {
+        var plyrInstance = document.getElementsByClassName("plyr")[0];
+        this.$nextTick(() => {
+          var maximizeButton = document.getElementById("maximizeButton");
+          if (maximizeButton != undefined)
+            plyrInstance.insertBefore(maximizeButton, plyrInstance.firstChild);
+        });
```
Comment (dalmia) [inline src/pages/Player.vue:348]: this same snippet is used below - can make a method?

File: src/pages/Player.vue [hunk 10]
```diff
@@ -26,21 +33,42 @@
...
-        @option-selected="optionSelected"
-      ></item-modal>
+      <!-- minimize button -->
+      <icon-button
+        v-if="isModalMinimized && showItemModal"
+        :titleConfig="minimizeButtonTitleConfig"
+        :buttonClass="minimizeButtonClass"
+        @click="toggleMinimize"
+        :class="[minimizeButtonPositionClass, 'absolute z-20 btn']"
+        id="maximizeButton"
```
Comment (dalmia) [inline src/pages/Player.vue:43]: why use minimize everywhere and maximize here?

**Comment by deepansh96** [comment]
- `justify-between`
 - padding right of `show question` button
 - show question button on the right of skip button

File: src/assets/locales/en.js [hunk 11]
```diff
@@ -30,7 +30,7 @@ export default {
       },
       add_question: "Add a question",
       show_video: "Show Video",
-      show_item: "Show Item",
+      show_question: "Show Question",
```
Comment (dalmia) [inline src/assets/locales/en.js:33]: ideally this should be done based on item type but it's okay for now.

File: src/pages/Editor.vue [hunk 12]
```diff
@@ -40,14 +40,16 @@
...
-                @click="toggleMinimize"
-                class="absolute z-20 top-2 right-2 btn"
-              >
-              </icon-button>
+              <transition name="maximize-btn-transition">
+                <icon-button
+                  v-if="showItemModal && isModalMinimized"
+                  :titleConfig="maximizeButtonTitleClass"
+                  :buttonClass="maximizeButtonClass"
+                  @click="maximize"
```
Comment (dalmia) [inline src/pages/Editor.vue:48]: a slightly more descriptive name of the function would be better: `maximizeModal`

File: src/pages/Editor.vue [hunk 13]
```diff
@@ -63,6 +65,7 @@
                   :selectedItemIndex="currentItemIndex"
                   :itemList="items"
                   :previewMode="true"
+                  @toggle-minimize="minimize"
```
Comment (dalmia) [inline src/pages/Editor.vue:64]: `minimizeModal`

File: src/assets/locales/hi.js [hunk 14]
```diff
@@ -30,7 +30,7 @@ export default {
       },
       add_question: "नया सवाल बनाएं",
       show_video: "वीडियो देखें",
-      show_item: "वस्तु देखें",
+      show_item: "सवाल देखें",
```
Comment (dalmia) [inline src/assets/locales/hi.js:33]: show question

File: src/pages/Editor.vue [hunk 15]
```diff
@@ -573,8 +581,21 @@ export default {
...
+    minimize(positions) {
+      // invoked when minimize button is clicked
+
+      // set some CSS variables which tells the animation
+      // where the modal should shrink to and where the maximize button should pop up
+      let root = document.documentElement;
+      root.style.setProperty("--t-origin-x", positions.centerX + "px");
+      root.style.setProperty("--t-origin-y", positions.centerY + "px");
+      root.style.setProperty("--maximize-btn-left", positions.leftX + "px");
+      root.style.setProperty("--maximize-btn-top", positions.leftY + "px");
```
Comment (dalmia) [inline src/pages/Editor.vue:589]: where are these 2 being used?

File: src/pages/Player.vue [hunk 16]
```diff
@@ -336,17 +336,30 @@ export default {
...
+      this.isModalMinimized = false;
+    },
+    minimize(positions) {
+      // invoked when minimize button is clicked
+
+      // set some CSS variables which tells the animation
+      // where the modal should shrink to and where the maximize button should pop up
+      let root = document.documentElement;
+      root.style.setProperty("--t-origin-x", positions.centerX + "px");
+      root.style.setProperty("--t-origin-y", positions.centerY + "px");
```
Comment (dalmia) [inline src/pages/Player.vue:351]: where are these 2 being used?

File: src/components/Items/Question/Header.vue [hunk 17]
```diff
@@ -1,19 +1,19 @@
 <template>
-  <div class="flex w-full bg-white justify-end p-1 pr-3 md:pr-6 space-x-2">
-    <!-- minimize button -->
-    <icon-button
-      :titleConfig="minimizeButtonTitleConfig"
-      :buttonClass="minimizeButtonClass"
-      @click="toggleMinimize"
-      id="minimize"
-    ></icon-button>
+  <div :class="mainStylingClass">
```
Comment (dalmia) [inline src/components/Items/Question/Header.vue:2]: better name for the class - `container` or something more standard.

File: src/components/Items/Question/Footer.vue [hunk 18]
```diff
@@ -1,5 +1,5 @@
 <template>
-  <div class="flex w-full bg-white p-1 py-2 md:p-3 justify-around place-self-end">
+  <div :class="mainStylingClass">
```
Comment (dalmia) [inline src/components/Items/Question/Footer.vue:2]: same comment on class name

**Comment by dalmia** [comment]
add question moves a bit out on portrait mode: (apart from the video I sent on Discord) image

**Pull request merged**